src/runtime/hexagon/rpc/hexagon/rpc_server.cc

```diff
@@ -0,0 +1,212 @@
+#include <AEEStdErr.h>
+#include <HAP_farf.h>
+#include <HAP_perf.h>
+#include <qurt_error.h>
+#include <qurt_hvx.h>
+}
+
+#include <dlfcn.h>
+#include <tvm/runtime/object.h>
+#include <tvm/runtime/packed_func.h>
+#include <tvm/runtime/registry.h>
+
+#include <algorithm>
+#include <memory>
+#include <string>
+
+#include "../../../library_module.h"
+#include "../../../minrpc/minrpc_server.h"
+#include "../../hexagon/hexagon_common.h"
+#include "hexagon_rpc.h"
+
+#define TVM_HEXAGON_RPC_BUFF_SIZE_BYTES 1024 * 1024
+
+#define TVM_LOG_CUSTOMIZE 1
+
+namespace tvm {
+namespace runtime {
+namespace hexagon {
+
+class HexagonIOHandler {
+ public:
+  explicit HexagonIOHandler(uint8_t* read_buffer) : read_buffer_{read_buffer} {}
+
+  void MessageStart(size_t message_size_bytes) {}
+
+  ssize_t PosixWrite(const uint8_t* buf, size_t write_len_bytes) {
+    FARF(ALWAYS, "HexagonIOHandler PosixWrite called, write_len_bytes: %d", write_len_bytes);
+    size_t written_size = static_cast<size_t>(
+        write_buffer_.sputn(reinterpret_cast<const char*>(buf), write_len_bytes));
+    if (written_size != write_len_bytes) {
+      FARF(ALWAYS, "HexagonIOHandler written_size failed");
+    }
+    return (ssize_t)written_size;
+  }
+
+  void MessageDone() {}
+
+  ssize_t PosixRead(uint8_t* buf, size_t read_len_bytes) {
+    FARF(ALWAYS, "HexagonIOHandler PosixRead called, %d, %d", read_len_bytes,
+         read_buffer_size_bytes_);
+
+    uint32_t bytes_to_read = 0;
+    if ((read_buffer_size_bytes_ - read_len_bytes) < 0) {
+      bytes_to_read = read_buffer_size_bytes_;
+    } else {
+      bytes_to_read = read_len_bytes;
+    }
+
+    std::memcpy(buf, read_buffer_, bytes_to_read);
+    read_buffer_ += bytes_to_read;
+    read_buffer_size_bytes_ -= bytes_to_read;
+    if (bytes_to_read != read_len_bytes) {
+      FARF(ERROR, "Error bytes_to_read (%d) < read_len_bytes (%d).", bytes_to_read, read_len_bytes);
+    }
+    return (ssize_t)bytes_to_read;
+  }
+
+  void SetReadBuffer(const uint8_t* buf, size_t buf_size_bytes) {
+    FARF(ALWAYS, "HexagonIOHandler SetReadBuffer called: %d, prev read_buffer_size_bytes_: ",
+         buf_size_bytes, read_buffer_size_bytes_);
+    read_buffer_ = buf;
+    read_buffer_size_bytes_ = buf_size_bytes;
+  }
+
+  int64_t GetWriteBuffer(uint8_t* buf, size_t read_len_bytes) {
+    FARF(ALWAYS, "HexagonIOHandler GetWriteBuffer called, read_len_bytes: %d", read_len_bytes);
+    return write_buffer_.sgetn(reinterpret_cast<char*>(buf), read_len_bytes);
+  }
+
+  void Close() { FARF(ALWAYS, "HexagonIOHandler Close called"); }
+
+  void Exit(int code) { exit(code); }
+
+ private:
+  const uint8_t* read_buffer_;
+  uint32_t read_buffer_size_bytes_;
+
+  std::stringbuf write_buffer_;
+};
+
+class HexagonRPCServer {
+ public:
+  explicit HexagonRPCServer(uint8_t* receive_buffer) : io_{receive_buffer}, rpc_server_{&io_} {};
+
+  int64_t Write(const uint8_t* data, size_t data_len_bytes) {
+    io_.SetReadBuffer(data, data_len_bytes);
```

Comment:
The names here are a bit confusing. Write calls SetReadBuffer, and below Read calls GetWriteBuffer.. I think I understand that HexagonRPCServer::Write sets the buffer to "read from" for the IO handler. 

Maybe some docs or more intuitive naming could help demystify.

Reply:
agreed, it is really confusing. I added some documentation here: https://github.com/apache/tvm/pull/9631/commits/e57c7aafcc650a644e5d93d4ba7145725c6cd538

Please let me know if you have more suggestions.